src/file-handler/index.ts

```diff
@@ -1,49 +1,133 @@
+                return this.loadToml(filename);
+            default:
+                throw new Error(`Invalid extension: ${ext}`);
+        }
+    }
+
+    loadYaml(filename: string): ConfigVersionDocument {
+        const file = readFileSync(filename, 'utf-8');
+        return parse(file);
+    }
+
+    loadToml(filename: string): ConfigVersionDocument {
+        // enable parsing of multiline inline tables
+        const opts = {
+            x: {
+                multi: true,
+            },
+        };
+
+        const file = readFileSync(filename, 'utf-8');
+
+        // there is some strangeness that needs further investigation
+        // where what comes out of TOML.parse, whilst serializing to an
+        // identical JSON object as yaml.load, behaves differently when
+        // subjected to lodash.merge. the workaround is to serialize
+        // to JSON and then parse back to an object
+        return JSON.parse(JSON.stringify(TOML.parse(file, opts))) as unknown as ConfigVersionDocument;
+    }
+
+    buildTree(): ConfigFileTree {
+        if (!this.opts) {
+            throw new Error('Input and output directories must be specified.')
+        }
+
+        const opts = this.opts
+
+        interface MiniTree {
+            [dir: string]: string[]
+        }
+
+        // recursively look through `dir` for .`ext` files
+        function findFilesWithExt(dir: string, ext: string) {
+            const output: MiniTree = {};
+            const files = readdirSync(dir);
+            for (const file of files) {
+                const filePath = path.join(dir, file);
+                const fileStat = statSync(filePath);
+                if (fileStat.isDirectory()) {
+                    merge(output, findFilesWithExt(filePath, ext));
+                } else if (file.endsWith(ext)) {
+                    const [directory, file] = sanitisePath(filePath);
+                    if (!output[directory]) {
+                        output[directory] = [];
+                    }
+                    output[directory].push(file);
+                }
+            }
+            return output;
+        }
+
+        function sanitisePath(filePath: string): [string, string] {
+            const relative = path.relative(opts.inputDirectory, filePath)
+            const directory = path.dirname(relative)
+            const file = path.basename(relative)
+            return [directory, file];
+        }
+
+        function findLatest(tree: MiniTree) {
+            const output: ConfigFileTree = {};
+            Object.keys(tree).forEach(dir => {
+                output[dir] = tree[dir].sort(sortBySemVer)[0]
+            })
+            return output;
+        }
+
+        function sortBySemVer(a: string, b: string): number {
+            if (opts.fileExtension) {
+                a = a.replace(opts.fileExtension, "")
+                b = b.replace(opts.fileExtension, "")
+            }
+            return new SemVer(b).compare(new SemVer(a));
+        }
```

Comment:
Just checking on this - it's comparing strings, so `1.2.3` compared with `1.2.3` will return true, but `1.2.3` compared with `2.2.3` will return false. However, it's not checking that `2.2.3` is later than `1.2.3`. I guess the question is do we ever need to do that?

Reply:
i _think_ this does what we need to enable sorting of filenames based on SemVer, newest first:
```
❯ node
Welcome to Node.js v16.20.2.
Type ".help" for more information.
> const {SemVer} = require('semver')
undefined
> new SemVer('1.2.3').compare(new SemVer('2.2.3'))
-1
> new SemVer('1.2.3').compare(new SemVer('1.2.3'))
0
> new SemVer('2.2.3').compare(new SemVer('1.2.3'))
1
```
based on the code [here](https://github.com/npm/node-semver/blob/0a12d6c7debb1dc82d8645c770e77c47bac5e1ea/classes/semver.js#L91) it should be doing it 'properly' rather than a string comparison (otherwise you'd get some awful stuff like 11.0.0 being 'less than' 2.0.0)

Reply:
Interesting! I wonder if that's TF kicking in and auto-casting types 🤔